File: backend/app/models/order.py
"""
Order models - заявки клиентов
"""
from sqlalchemy import Column, Integer, String, Numeric, DateTime, ForeignKey, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.database import Base
import enum


class OrderStatus(str, enum.Enum):
    """Статусы заявки"""
    DRAFT = "draft"
    AWAITING_PAYMENT = "awaiting_payment"
    PAID = "paid"
    IN_PROGRESS = "in_progress"
    READY = "ready"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


class Order(Base):
    """Модель заявки"""
    __tablename__ = "orders"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    service_id = Column(Integer, ForeignKey("services.id"), nullable=False)
    params = Column(JSONB, nullable=False)
    estimate_total = Column(Numeric(15, 2), nullable=False)
    status = Column(SQLEnum(OrderStatus), default=OrderStatus.DRAFT, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    deadline = Column(DateTime(timezone=True), nullable=True)
    document_url = Column(String(500), nullable=True)
    
    # Relationships
    user = relationship("User", back_populates="orders")
    service = relationship("Service", back_populates="orders")
    payments = relationship("Payment", back_populates="order")
    status_history = relationship("OrderStatusHistory", back_populates="order")


class OrderStatusHistory(Base):
    """История изменений статусов заявки"""
    __tablename__ = "order_status_history"
    
    id = Column(Integer, primary_key=True, index=True)
    order_id = Column(Integer, ForeignKey("orders.id"), nullable=False, index=True)
    old_status = Column(String(30), nullable=False)
    new_status = Column(String(30), nullable=False)
    actor_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    comment = Column(String(500), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    order = relationship("Order", back_populates="status_history")

File: frontend/lib/models/order.dart
class Order {
  final int id;
  final int userId;
  final int serviceId;
  final Map<String, dynamic> params;
  final double estimateTotal;
  final String status;
  final DateTime createdAt;
  final DateTime? deadline;
  final String? documentUrl;
  
  Order({
    required this.id,
    required this.userId,
    required this.serviceId,
    required this.params,
    required this.estimateTotal,
    required this.status,
    required this.createdAt,
    this.deadline,
    this.documentUrl,
  });
  
  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'],
      userId: json['user_id'],
      serviceId: json['service_id'],
      params: json['params'],
      estimateTotal: (json['estimate_total'] as num).toDouble(),
      status: json['status'],
      createdAt: DateTime.parse(json['created_at']),
      deadline: json['deadline'] != null ? DateTime.parse(json['deadline']) : null,
      documentUrl: json['document_url'],
    );
  }
  
  /// Получить читаемое название статуса на русском
  String get statusRu {
    switch (status) {
      case 'draft':
        return 'Черновик';
      case 'awaiting_payment':
        return 'Ожидает оплаты';
      case 'paid':
        return 'Оплачено';
      case 'in_progress':
        return 'В работе';
      case 'ready':
        return 'Готово';
      case 'delivered':
        return 'Выдан';
      case 'cancelled':
        return 'Отменён';
      default:
        return status;
    }
  }
}
